Search

phuedx opened the pull request:
### Notes
- Requires joakin/loot#37
### TODO
- [x] Focus search bar when the search overlay opens
- [x] Add a close button to the search overlay
- [x] Handle the user typing quickly, e.g. by debouncing the `onChange` event listener or by ignoring stale `STOP_SEARCH_REQUEST` actions in the `search` reducer
- [x] Link to article on search results

phuedx committed:
Add basic chrome and search box

Copypasta responsive breakpoints for the Article component for now with
a view to defining our style sharing mechanism(s) later.

phuedx committed:
Add the Search container

Add a container that will handle the user's interactions with the search
bar. For now, dispatch the START_SEARCH_SESSION action when the user
clicks the search bar.

Additional changes:
* Fix standard errors in the Search component

phuedx committed:
Show search overlay when user clicks search bar

Changes:
* Add search.isSearching to the global state, which defaults to false,
  and tracks whether the user is in a search session
* If the user is in a search session, then render the SearchOverlay
  component

phuedx committed:
Add the search API

Following the pattern of the lead and slim APIs, proxy requests to
/api/search/[q] to http://loot/search/[q].

phuedx committed:
Dispatch search when user types three charaters

When the user types three characters or more, dispatch: the
START_SEARCH_REQUEST action, the API request itself, and, when the API
request stops, the STOP_SEARCH_REQUEST action.

phuedx committed:
Display search results

phuedx committed:
Focus search bar when search overlay opens

Changes:
* Add the HTML5 autofocus attribute to the search bar
* Always set the attribute to truthy when the search overlay is rendered

phuedx committed:
Ignore search results from stale searches

joakin commented:
Tested, works, missing links to articles in results. Gonna add it to the PR description as a TODO

phuedx committed:
[CR] Add the .content-container mixin

Share the basic responsive breakpoints for content containers by adding
the .content-container mixin and using it in the .Article,
.Chrome-content, and .SearchOverlayResults rules.

Additional changes:
* Fix standard errors in the SearchOverlay component

phuedx committed:
[CR] Key search results on their title

Per https://facebook.github.io/react/docs/multiple-components.html#dynamic-children

Additional changes:
* Add TODO about extracting PageList component

phuedx committed:
[CR] Clear state when search overlay opens

phuedx committed:
[CR] Return search API call from search action

joakin commented:
Nice, I'm submitting a whitespace fix commit.

Looking good!

joakin committed:
[CR] Whitespace fix

phuedx committed:
Make search results link to articles

phuedx committed:
Add close button to the search overlay

Changes:
* Add a close button to the Chrome component, which isn't rendered by default
* Render the close button in the search overlay
* When the user clicks the button, close the search overlay by dispatching the STOP_SEARCH_SESSION
  action

phuedx commented:
**TODONE** 🎉
